File: app/src/main/java/com/mittylabs/elaps/ui/main/TimerRunningActivity.java
//package com.mittylabs.elaps.ui.main;
//
//import android.app.Activity;
//import android.app.ActivityManager;
//import android.app.NotificationManager;
//import android.content.BroadcastReceiver;
//import android.content.Context;
//import android.content.Intent;
//import android.content.IntentFilter;
//import android.graphics.PorterDuff;
//import android.graphics.drawable.Drawable;
//import android.os.Build;
//import android.os.Bundle;
//import android.view.MenuItem;
//import android.view.View;
//import android.view.Window;
//import android.view.WindowManager;
//import android.widget.ImageView;
//import android.widget.ProgressBar;
//import android.widget.TextView;
//import android.widget.Toolbar;
//
//import androidx.annotation.Nullable;
//
//import com.mittylabs.elaps.R;
//
//import java.util.concurrent.TimeUnit;
//
//public class TimerRunningActivity extends Activity implements View.OnClickListener {
//
//    private ProgressBar progressCircle;
//    private TextView textViewTime;
//    private ImageView imageViewStartStop;
//    private Toolbar toolbar;
//    private TimerSetting timerSetting;
//
//    public static Intent getLaunchingIntent(final Context ctx) {
//        return new Intent(ctx, TimerRunningActivity.class);
//    }
//
//    @Override
//    protected void onCreate(@Nullable Bundle savedInstanceState) {
//        super.onCreate(savedInstanceState);
//        setContentView(R.layout.activity_timer_running);
//        setStatusBarColor(R.color.color_primary);
//
//        realm = Realm.getDefaultInstance();
//        timerSetting = realm.where(TimerSetting.class).equalTo("id", Constants.DEFAULT_TIMER_ID).findFirst();
//
//        progressCircle = (ProgressBar) findViewById(R.id.barTimer);
//        ImageView imageViewReset = (ImageView) findViewById(R.id.timer_reset_button);
//        textViewTime = (TextView) findViewById(R.id.textViewTime);
//        imageViewStartStop = (ImageView) findViewById(R.id.timer_start_button);
//        ImageView stopButton = (ImageView) findViewById(R.id.timer_stop_button);
//        toolbar = (Toolbar) findViewById(R.id.toolbar_timer);
//        setupToolbar();
//
//        imageViewReset.setOnClickListener(this);
//        imageViewStartStop.setOnClickListener(this);
//        stopButton.setOnClickListener(this);
//
//        resetProgressBar(timerSetting.getInitialTimeMilliseconds());
//        startTimerService();
//    }
//
//    private final BroadcastReceiver br = new BroadcastReceiver() {
//        @Override
//        public void onReceive(Context context, Intent intent) {
//            receiveBroadcast(intent);
//        }
//    };
//
//    /**
//     * register listening broadcast receiver to get timer onTick callbacks
//     */
//    @Override
//    public void onResume() {
//        super.onResume();
//        registerReceiver(br, new IntentFilter(TimerBroadcastService.COUNTDOWN_BR));
//    }
//
//    /**
//     * unregister listening broadcast receiver to stop getting timer onTick callbacks
//     */
//    @Override
//    public void onPause() {
//        super.onPause();
//        unregisterReceiver(br);
//    }
//
//    @Override
//    public void onStop() {
//        try {
//            unregisterReceiver(br);
//        } catch (Exception e) {
//            // Receiver was probably already stopped in onPause()
//        }
//        super.onStop();
//    }
//
//    @Override
//    protected void onDestroy() {
//        realm.close();
//        super.onDestroy();
//    }
//
//    /**
//     * Handles the received broadcasts from the TimerBroadcastService
//     * @param intent received from broadcast service
//     */
//    private void receiveBroadcast(Intent intent) {
//        if (intent.getExtras() != null) {
//            final long millisUntilFinished = intent.getLongExtra("countdown", 0);
//            boolean finished = intent.getBooleanExtra("finished", false);
//
//            /** If timer is not yet finished, update the time and progress bar */
//            if (!finished) {
//                setProgressBar(millisUntilFinished);
//                String formattedTime = formatTime(millisUntilFinished);
//                textViewTime.setText(formattedTime);
//            }
//            /** If timer is finished, reset the timer */
//            else {
//                resetTimer();
//                imageViewStartStop.setImageResource(R.drawable.icon_play_orange);
//            }
//        }
//    }
//
//    @Override
//    public void onClick(View view) {
//        switch (view.getId()) {
//            case R.id.timer_reset_button:
//                resetTimer();
//                break;
//            case R.id.timer_start_button:
//                timerStatusToggle(timerSetting.getStatus());
//                break;
//            case R.id.timer_stop_button:
//                resetTimer();
//                finish();
//                break;
//        }
//    }
//
//    /**
//     * - reset the displayed time to the initial time
//     * - reset the saved realm time to the initial time and set the timer status to STOP
//     * - reset the progress bar
//     * - stop the TimerBroadcastService
//     */
//    private void resetTimer() {
//        String formattedTime = formatTime(timerSetting.getInitialTimeMilliseconds());
//        textViewTime.setText(formattedTime);
//        realm.executeTransaction(new Realm.Transaction() {
//            @Override
//            public void execute(@Nonnull Realm realm) {
//                timerSetting.setSavedTimeMilliseconds(timerSetting.getInitialTimeMilliseconds());
//                timerSetting.setStatus(TimerSetting.Status.STOPPED);
//            }
//        });
//        resetProgressBar(timerSetting.getInitialTimeMilliseconds());
//        imageViewStartStop.setImageResource(R.drawable.icon_play_orange);
//        NotificationManager notificationManager = (NotificationManager)getSystemService(Context.NOTIFICATION_SERVICE);
//        notificationManager.cancel(NOTIFICATION_STICKY_ID);
//        stopTimerService();
//    }
//
//    private void resetProgressBar(long time) {
//        int max = (int) time / 1000;
//        progressCircle.setMax(max);
//        progressCircle.setProgress(max);
//    }
//
//    private void setProgressBar(long time) {
//        int progress = (int) (time / 1000);
//        progressCircle.setProgress(progress);
//    }
//
//    public static String formatTime(long time) {
//        return String.format("%02d:%02d:%02d",
//                TimeUnit.MILLISECONDS.toHours(time),
//                TimeUnit.MILLISECONDS.toMinutes(time) - TimeUnit.HOURS.toMinutes(TimeUnit.MILLISECONDS.toHours(time)),
//                TimeUnit.MILLISECONDS.toSeconds(time) - TimeUnit.MINUTES.toSeconds(TimeUnit.MILLISECONDS.toMinutes(time)));
//    }
//
//    /**
//     * handles the interactions to play/pause
//     * @param timerStatus indicates the current status
//     */
//    private void timerStatusToggle(TimerSetting.Status timerStatus) {
//        switch (timerStatus) {
//            case STOPPED: // press play to start
//                resetProgressBar(timerSetting.getInitialTimeMilliseconds());
//                imageViewStartStop.setImageResource(R.drawable.icon_pause_orange);
//                realm.executeTransaction(new Realm.Transaction() {
//                    @Override
//                    public void execute(@Nonnull Realm realm) {
//                        timerSetting.setStatus(TimerSetting.Status.STARTED);
//                    }
//                });
//                startTimerService();
//                break;
//            case STARTED: // press pause to pause
//                imageViewStartStop.setImageResource(R.drawable.icon_play_orange);
//                realm.executeTransaction(new Realm.Transaction() {
//                    @Override
//                    public void execute(@Nonnull Realm realm) {
//                        timerSetting.setStatus(TimerSetting.Status.PAUSED);
//                    }
//                });
//                NotificationUtilts.alowNotificationToBeRemoved("Time left: " + formatTime(timerSetting.getSavedTimeMilliseconds()) , this);
//                stopTimerService();
//                break;
//            case PAUSED: // unpause
//                imageViewStartStop.setImageResource(R.drawable.icon_pause_orange);
//                realm.executeTransaction(new Realm.Transaction() {
//                    @Override
//                    public void execute(@Nonnull Realm realm) {
//                        timerSetting.setStatus(TimerSetting.Status.STARTED);
//                    }
//                });
//                startTimerService();
//                break;
//        }
//    }
//
//    /**
//     * Start a new TimerBroadcast Service to persist the timer countdown
//     */
//    private void startTimerService() {
//        if (!isMyServiceRunning(TimerBroadcastService.class)) {
//            realm.executeTransaction(new Realm.Transaction() {
//                @Override
//                public void execute(@Nonnull Realm realm) {
//                    timerSetting.setStatus(TimerSetting.Status.STARTED); // update the status
//                }
//            });
//            Intent serviceIntent = new Intent(this, TimerBroadcastService.class);
//            serviceIntent.putExtra("TimerValue", timerSetting.getSavedTimeMilliseconds());
//            startService(serviceIntent);
//        }
//    }
//
//    /**
//     * Stop the TimerBroadcast Service if it's running
//     */
//    private void stopTimerService() {
//        if (isMyServiceRunning(TimerBroadcastService.class)) {
//            stopService(new Intent(this, TimerBroadcastService.class));
//        }
//    }
//
//    /**
//     * check if a service is currently running
//     * @param serviceClass is always TimerBroadcastService.class
//     * @return true if the service is running
//     */
//    private boolean isMyServiceRunning(Class<?> serviceClass) {
//        ActivityManager manager = (ActivityManager) getSystemService(Context.ACTIVITY_SERVICE);
//        for (ActivityManager.RunningServiceInfo service : manager.getRunningServices(Integer.MAX_VALUE)) {
//            if (serviceClass.getName().equals(service.service.getClassName())) {
//                return true;
//            }
//        }
//        return false;
//    }
//
//    private void setupToolbar() {
//        toolbar.setTitle("Timer");
//        setSupportActionBar(toolbar);
//        getSupportActionBar().setHomeButtonEnabled(true);
//        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
//        final Drawable downArrow = ContextCompat.getDrawable(this, R.drawable.icon_chevrondown);
//        downArrow.setColorFilter(ContextCompat.getColor(this, R.color.white), PorterDuff.Mode.SRC_ATOP);
//        getSupportActionBar().setHomeAsUpIndicator(downArrow);
//    }
//
//    private void setStatusBarColor(int color) {
//            Window window = this.getWindow();
//            window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
//            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
//            window.setStatusBarColor(ContextCompat.getColor(this, color));
//    }
//
//    @Override
//    public boolean onOptionsItemSelected(MenuItem item) {
//        switch (item.getItemId()) {
//            case android.R.id.home:
//                onBackPressed();
//                break;
//            default:
//                break;
//        }
//        return super.onOptionsItemSelected(item);
//    }
//
//    /**
//     * @return a unique constant ID to save the state of the presenters
//     */
//    @Override
//    public int getViewId() {
//        return Constants.ACTIVITY_TIMER_ID;
//    }
//}

File: app/src/main/java/com/mittylabs/elaps/ui/main/TimerRunningActivity.kt
package com.mittylabs.elaps.ui.main

import android.app.Activity
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.view.MenuItem
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.mittylabs.elaps.databinding.ActivityTimerRunningBinding
import com.mittylabs.elaps.service.CountdownTimerService
import com.mittylabs.elaps.service.CountdownTimerService.Companion.INTENT_EXTRA_INITIAL_TIME
import com.mittylabs.elaps.service.CountdownTimerService.Companion.INTENT_EXTRA_RESULT
import com.mittylabs.elaps.ui.main.TimerSetupActivity.Companion.INTENT_EXTRA_MINUTES
import com.mittylabs.elaps.ui.toTimerFormat


class TimerRunningActivity : Activity() {
    private lateinit var receiver: TimerStatusReceiver
    private lateinit var binding: ActivityTimerRunningBinding

    private val timerLengthMinutes by lazy {
        minToMillis(intent.getIntExtra(INTENT_EXTRA_MINUTES, 0))
    }
    private var timerState = TimerState.STOPPED

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityTimerRunningBinding.inflate(layoutInflater)
        setContentView(binding.root)

        receiver = TimerStatusReceiver()

        binding.timerResetButton.setOnClickListener {
            resetProgressBar()
        }
        binding.timerStartButton.setOnClickListener {
            startService()
        }
        binding.timerStopButton.setOnClickListener {
            stopService()
        }
    }

    override fun onResume() {
        super.onResume()

        LocalBroadcastManager.getInstance(this)
            .registerReceiver(receiver, IntentFilter(CountdownTimerService.TIME_INFO))
    }

    override fun onPause() {
        super.onPause()
        LocalBroadcastManager.getInstance(this).unregisterReceiver(receiver)
    }

    private fun startService() {
        val intent = Intent(this, CountdownTimerService::class.java)
        intent.putExtra(INTENT_EXTRA_MINUTES, timerLengthMinutes)
        startService(intent)
    }

    private fun stopService() {
        val intent = Intent(this, CountdownTimerService::class.java)
        stopService(intent)
    }

    inner class TimerStatusReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (intent.action == CountdownTimerService.TIME_INFO) {
                if (intent.hasExtra(INTENT_EXTRA_RESULT)) {
                    val time = intent.getLongExtra(INTENT_EXTRA_RESULT, 0L)
                    val initialTime = intent.getLongExtra(INTENT_EXTRA_INITIAL_TIME, 0L)
                    updateProgressBar(initialTime, time)
                    binding.textViewTime.text = time.toTimerFormat()
                }
            }
        }
    }

//    /**
//     * - reset the displayed time to the initial time
//     * - reset the saved realm time to the initial time and set the timer status to STOP
//     * - reset the progress bar
//     * - stop the TimerBroadcastService
//     */
//    private fun resetTimer() {
//        val formattedTime = formatTime(timerSetting!!.initialTimeMilliseconds)
//        binding.textViewTime!!.text = formattedTime
//        realm.executeTransaction(object : Transaction() {
//            fun execute(@Nonnull realm: Realm?) {
//                timerSetting!!.savedTimeMilliseconds = timerSetting!!.initialTimeMilliseconds
//                timerSetting!!.status = TimerSetting.Status.STOPPED
//            }
//        })
//        resetProgressBar(timerSetting!!.initialTimeMilliseconds)
//        binding.timerStartButton!!.setImageResource(R.drawable.icon_play_orange)
//        val notificationManager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
//        notificationManager.cancel(NOTIFICATION_STICKY_ID)
//        stopTimerService()
//    }

    private fun resetProgressBar() {
        binding.barTimer.max = timerLengthMinutes.toInt()
        binding.barTimer.progress = 0
    }

    private fun updateProgressBar(initialTime: Long, time: Long) {
        binding.barTimer.max = initialTime.toInt()

        val result = (initialTime - time)
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            binding.barTimer.setProgress(result.toInt(), true)
        } else {
            binding.barTimer.progress = result.toInt()
        }
    }

//    /**
//     * handles the interactions to play/pause
//     * @param timerStatus indicates the current status
//     */
//    private fun timerStatusToggle(timerStatus: TimerSetting.Status) {
//        when (timerStatus) {
//            TimerSetting.Status.STOPPED -> {
//                resetProgressBar(timerSetting!!.initialTimeMilliseconds)
//                binding.timerStartButton!!.setImageResource(R.drawable.icon_pause_orange)
//                realm.executeTransaction(object : Transaction() {
//                    fun execute(@Nonnull realm: Realm?) {
//                        timerSetting!!.status = TimerSetting.Status.STARTED
//                    }
//                })
//                startTimerService()
//            }
//            TimerSetting.Status.STARTED -> {
//                binding.timerStartButton!!.setImageResource(R.drawable.icon_play_orange)
//                realm.executeTransaction(object : Transaction() {
//                    fun execute(@Nonnull realm: Realm?) {
//                        timerSetting!!.status = TimerSetting.Status.PAUSED
//                    }
//                })
//                NotificationUtilts.alowNotificationToBeRemoved(
//                    "Time left: " + formatTime(
//                        timerSetting!!.savedTimeMilliseconds
//                    ), this
//                )
//                stopTimerService()
//            }
//            TimerSetting.Status.PAUSED -> {
//                binding.timerStartButton!!.setImageResource(R.drawable.icon_pause_orange)
//                realm.executeTransaction(object : Transaction() {
//                    fun execute(@Nonnull realm: Realm?) {
//                        timerSetting!!.status = TimerSetting.Status.STARTED
//                    }
//                })
//                startTimerService()
//            }
//        }
//    }


    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> onBackPressed()
            else -> {
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun minToMillis(minutes: Int): Long = minutes * 60L * 1000L

    companion object {
        fun getLaunchingIntent(ctx: Context?): Intent {
            return Intent(ctx, TimerRunningActivity::class.java)
        }
//
//        fun setAlarm(context: Context, nowSeconds: Long, secondsRemaining: Long): Long {
//            val wakeUpTime = (nowSeconds + secondsRemaining) * 1000
//            val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
//            val intent = Intent(context, TimerExpiredReceiver::class.java)
//            val pendingIntent = PendingIntent.getBroadcast(context, 0, intent, 0)
//            alarmManager.setExact(AlarmManager.RTC_WAKEUP, wakeUpTime, pendingIntent)
//            PrefUtil.setAlarmSetTime(nowSeconds, context)
//            return wakeUpTime
//        }
//
//        fun removeAlarm(context: Context) {
//            val intent = Intent(context, TimerExpiredReceiver::class.java)
//            val pendingIntent = PendingIntent.getBroadcast(context, 0, intent, 0)
//            val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
//            alarmManager.cancel(pendingIntent)
//            PrefUtil.setAlarmSetTime(0, context)
//        }
    }
}